File: examples/RayMarcher.scala
//> using scala 3.3.7
//> using dep xyz.matthieucourt::layoutz:0.7.0

import layoutz._


case class V3(x: Double, y: Double, z: Double):
  def +(o: V3): V3     = V3(x + o.x, y + o.y, z + o.z)
  def -(o: V3): V3     = V3(x - o.x, y - o.y, z - o.z)
  def *(s: Double): V3  = V3(x * s, y * s, z * s)
  def dot(o: V3): Double = x * o.x + y * o.y + z * o.z
  def cross(o: V3): V3  = V3(y * o.z - z * o.y, z * o.x - x * o.z, x * o.y - y * o.x)
  def length: Double     = math.sqrt(this dot this)
  def normalized: V3     = { val l = length; if l < 1e-10 then V3.Zero else this * (1.0 / l) }
  def unary_- : V3      = V3(-x, -y, -z)

object V3:
  val Zero    = V3(0, 0, 0)
  val Up      = V3(0, 1, 0)
  val Forward = V3(0, 0, 1)


case class Pixel(ch: Char, r: Int, g: Int, b: Int)


case class RayState(
    theta: Double,
    phi: Double,
    dist: Double,
    morph: Double,
    morphTarget: Double,
    autoRotate: Boolean,
    tick: Int,
    frameNs: List[Long]
)

object RayState:
  def apply(): RayState =
    RayState(0.0, 0.35, 3.8, 0.0, 0.0, true, 0, List.fill(30)(80000000L))

sealed trait RayMsg
case object RTick extends RayMsg
case object RotL extends RayMsg
case object RotR extends RayMsg
case object RotU extends RayMsg
case object RotD extends RayMsg
case object ZoomIn extends RayMsg
case object ZoomOut extends RayMsg
case object ToggleAuto extends RayMsg
case object NextShape extends RayMsg

object RayMarcher extends LayoutzApp[RayState, RayMsg]:

  private val W = 60
  private val H = 28

  private val MaxSteps  = 50
  private val MaxDist   = 20.0
  private val Eps       = 0.005
  private val MorphSpeed = 0.06

  private val Ramp = " .'`^\",:;Il!i><~+_-?][}{1)(|/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$"
  private val ShapeNames = Vector("sphere", "torus", "cube")

  private def clamp(x: Double, lo: Double, hi: Double): Double = math.max(lo, math.min(hi, x))
  private def mix(a: Double, b: Double, t: Double): Double = a * (1.0 - t) + b * t
  private def smoothstep(lo: Double, hi: Double, x: Double): Double =
    val t = clamp((x - lo) / (hi - lo), 0.0, 1.0); t * t * (3.0 - 2.0 * t)

  private def sdSphere(p: V3, r: Double): Double =
    p.length - r

  private def sdTorus(p: V3, R: Double, r: Double): Double =
    val qx = math.sqrt(p.x * p.x + p.z * p.z) - R
    math.sqrt(qx * qx + p.y * p.y) - r

  private def sdRoundBox(p: V3, b: V3, r: Double): Double =
    val q = V3(math.abs(p.x) - b.x, math.abs(p.y) - b.y, math.abs(p.z) - b.z)
    val outer = V3(math.max(q.x, 0), math.max(q.y, 0), math.max(q.z, 0)).length
    outer + math.min(math.max(q.x, math.max(q.y, q.z)), 0.0) - r

  private def scene(p: V3, morph: Double): Double =
    val t      = ((morph % 3.0) + 3.0) % 3.0
    val sphere = sdSphere(p, 1.05)
    val torus  = sdTorus(p, 0.9, 0.38)
    val cube   = sdRoundBox(p, V3(0.72, 0.72, 0.72), 0.12)
    if t < 1.0      then mix(sphere, torus, smoothstep(0, 1, t))
    else if t < 2.0 then mix(torus, cube,   smoothstep(0, 1, t - 1))
    else                  mix(cube, sphere,  smoothstep(0, 1, t - 2))

  private def calcNormal(p: V3, morph: Double): V3 =
    val e = 0.001
    V3(
      scene(V3(p.x + e, p.y, p.z), morph) - scene(V3(p.x - e, p.y, p.z), morph),
      scene(V3(p.x, p.y + e, p.z), morph) - scene(V3(p.x, p.y - e, p.z), morph),
      scene(V3(p.x, p.y, p.z + e), morph) - scene(V3(p.x, p.y, p.z - e), morph)
    ).normalized

  @annotation.tailrec
  private def march(ro: V3, rd: V3, morph: Double, t: Double = 0.0, i: Int = 0): Double =
    if i >= MaxSteps || t >= MaxDist then -1.0
    else
      val d = scene(ro + rd * t, morph)
      if d < Eps then t else march(ro, rd, morph, t + d, i + 1)

  private val LightDir = V3(0.8, 1.0, -0.6).normalized

  private def shade(ro: V3, rd: V3, morph: Double): Pixel =
    val t = march(ro, rd, morph)
    if t < 0 then return bgPixel(rd)

    val hit  = ro + rd * t
    val n    = calcNormal(hit, morph)
    val diff = math.max(n dot LightDir, 0.0)
    val refl = n * (2.0 * (n dot LightDir)) - LightDir
    val spec = math.pow(math.max(refl dot -rd, 0.0), 32.0) * 0.6
    val ao   = 1.0 - clamp(scene(hit + n * 0.1, morph) * 5.0, 0.0, 0.4)
    val lum  = clamp((0.08 + diff * 0.85 + spec) * ao, 0.0, 1.0)

    val ch = Ramp(clamp(lum * (Ramp.length - 1), 0, Ramp.length - 1).toInt)

    val nx = n.x * 0.5 + 0.5; val ny = n.y * 0.5 + 0.5; val nz = n.z * 0.5 + 0.5
    Pixel(
      ch,
      clamp(( nx * 0.55 + lum * 0.45) * 235 + 20, 0, 255).toInt,
      clamp(( ny * 0.45 + lum * 0.55) * 215 + 15, 0, 255).toInt,
      clamp(( nz * 0.50 + lum * 0.50 + 0.05) * 200 + 30, 0, 255).toInt
    )

  private def bgPixel(rd: V3): Pixel =
    val vy  = rd.y * 0.5 + 0.5
    val bg  = (18.0 - (1.0 - vy) * 8.0).max(4.0).toInt
    Pixel(' ', bg, bg, bg + 6)

  case class FrameBuffer(pixels: Array[Pixel], w: Int, h: Int) extends Element:
    def render: String =
      val sb = new java.lang.StringBuilder(w * h * 20)
      var i = 0; var y = 0
      while y < h do
        var x = 0
        while x < w do
          val p = pixels(i)
          sb.append("\u001b[38;2;").append(p.r).append(';').append(p.g).append(';').append(p.b).append('m').append(p.ch)
          i += 1; x += 1
        sb.append("\u001b[0m")
        if y < h - 1 then sb.append('\n')
        y += 1
      sb.toString

  private def renderFrame(s: RayState): FrameBuffer =
    val ro    = V3(
      s.dist * math.sin(s.theta) * math.cos(s.phi),
      s.dist * math.sin(s.phi),
      s.dist * math.cos(s.theta) * math.cos(s.phi)
    )
    val fwd   = (-ro).normalized
    val right = (fwd cross V3.Up).normalized
    val up    = right cross fwd
    val aspect = W.toDouble / H.toDouble * 0.48

    val pixels = new Array[Pixel](W * H)
    var i = 0; var py = 0
    while py < H do
      val v = 0.5 - py.toDouble / H
      var px = 0
      while px < W do
        val u  = (px.toDouble / W - 0.5) * aspect
        val rd = (fwd + right * u + up * v).normalized
        pixels(i) = shade(ro, rd, s.morph)
        i += 1; px += 1
      py += 1
    FrameBuffer(pixels, W, H)

  def init: (RayState, Cmd[RayMsg]) =
    (RayState(), Cmd.setTitle("layoutz · Ray Marcher"))

  def update(msg: RayMsg, s: RayState): (RayState, Cmd[RayMsg]) = msg match
    case RTick =>
      val now      = System.nanoTime()
      val newTheta = if s.autoRotate then s.theta + 0.035 else s.theta
      val diff     = s.morphTarget - s.morph
      val newMorph = if math.abs(diff) < 0.01 then s.morphTarget else s.morph + diff * MorphSpeed
      val ft       = (s.frameNs :+ (System.nanoTime() - now)).takeRight(30)
      s.copy(theta = newTheta, morph = newMorph, tick = s.tick + 1, frameNs = ft)

    case RotL  => s.copy(theta = s.theta - 0.15, autoRotate = false)
    case RotR  => s.copy(theta = s.theta + 0.15, autoRotate = false)
    case RotU  => s.copy(phi = math.min(s.phi + 0.1, 1.3))
    case RotD  => s.copy(phi = math.max(s.phi - 0.1, -1.3))
    case ZoomIn  => s.copy(dist = math.max(s.dist - 0.25, 2.0))
    case ZoomOut => s.copy(dist = math.min(s.dist + 0.25, 8.0))
    case ToggleAuto => s.copy(autoRotate = !s.autoRotate)
    case NextShape  =>
      val next = (math.round(s.morphTarget).toInt + 1) % 3
      s.copy(morphTarget = next.toDouble)

  def subscriptions(s: RayState): Sub[RayMsg] = Sub.batch(
    Sub.time.everyMs(80, RTick),
    Sub.onKeyPress {
      case Key.Left      => Some(RotL)
      case Key.Right     => Some(RotR)
      case Key.Up        => Some(RotU)
      case Key.Down      => Some(RotD)
      case Key.Char('+') => Some(ZoomIn)
      case Key.Char('-') => Some(ZoomOut)
      case Key.Char('a') => Some(ToggleAuto)
      case Key.Char('m') => Some(NextShape)
      case _             => None
    }
  )

  def view(s: RayState): Element =
    val fb        = renderFrame(s)
    val targetIdx = (math.round(s.morphTarget).toInt % 3 + 3) % 3
    val settled   = math.abs(s.morph - s.morphTarget) < 0.02
    val shapeName =
      if settled then ShapeNames(targetIdx)
      else
        val fromIdx = (math.round(s.morph).toInt % 3 + 3) % 3
        s"${ShapeNames(fromIdx)} → ${ShapeNames(targetIdx)}"

    val sparkData = s.frameNs.map(ns => (ns / 1e6).max(0.1))

    columns(
      layout(
        "Ray Marcher".style(Style.Bold).color(Color.BrightCyan),
        fb
      ),
      box("Camera")(
        leftAlign(layout(
          kv(
            "θ" -> f"${s.theta % (2 * math.Pi)}%.2f",
            "φ" -> f"${s.phi}%.2f",
            "zoom" -> f"${s.dist}%.1f"
          ).color(Color.BrightBlue),
          br,
          rowTight(
            "rotate: ".color(Color.BrightBlack),
            if s.autoRotate then "auto".color(Color.BrightGreen).style(Style.Bold)
            else "manual".color(Color.BrightYellow)
          ),
          br,
          s"~12 fps".color(Color.BrightYellow),
          sparkline(sparkData).color(Color.BrightCyan),
          br,
          kv("shape" -> shapeName).color(Color.BrightMagenta),
          br,
          spinner("render", s.tick / 2, SpinnerStyle.Dots).color(Color.BrightCyan),
          spinner("scene", s.tick / 3, SpinnerStyle.Earth).color(Color.BrightGreen),
          spinner("light", s.tick / 2, SpinnerStyle.Moon).color(Color.BrightYellow),
          br,
          layout(
            "←→↑↓  orbit".color(Color.BrightYellow),
            "+/-    zoom".color(Color.BrightYellow),
            "m      shape".color(Color.BrightYellow),
            "a      auto".color(Color.BrightYellow)
          ).style(Style.Dim)
        ), 22)
      ).border(Border.Round).color(Color.BrightMagenta)
    )

object RayMarcherRunner:
  def main(args: Array[String]): Unit =
    RayMarcher.run(quitKey = Key.Ctrl('Q'))

File: layoutz-hs/examples/RayMarcher.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE LambdaCase #-}

module Main where

import Layoutz
import Data.List (intercalate)
import Text.Printf (printf)

-- | 3D vector
data V3 = V3 !Double !Double !Double

vzero, vup :: V3
vzero = V3 0 0 0
vup   = V3 0 1 0

vadd, vsub, vcross :: V3 -> V3 -> V3
vadd (V3 ax ay az) (V3 bx by bz) = V3 (ax + bx) (ay + by) (az + bz)
vsub (V3 ax ay az) (V3 bx by bz) = V3 (ax - bx) (ay - by) (az - bz)
vcross (V3 ax ay az) (V3 bx by bz) =
  V3 (ay * bz - az * by) (az * bx - ax * bz) (ax * by - ay * bx)

vscale :: V3 -> Double -> V3
vscale (V3 x y z) s = V3 (x * s) (y * s) (z * s)

vdot :: V3 -> V3 -> Double
vdot (V3 ax ay az) (V3 bx by bz) = ax * bx + ay * by + az * bz

vlen :: V3 -> Double
vlen v = sqrt (vdot v v)

vneg :: V3 -> V3
vneg (V3 x y z) = V3 (-x) (-y) (-z)

vnorm :: V3 -> V3
vnorm v =
  let l = vlen v
  in if l < 1e-10 then vzero else vscale v (1.0 / l)

-- | One screen cell with character + 24-bit color
data Pixel = Pixel !Char !Int !Int !Int

-- | Render a list of pixels into a single ANSI string of width @w@
renderFrameBuffer :: [Pixel] -> Int -> String
renderFrameBuffer pixels w = intercalate "\n" (rowsOf pixels)
  where
    rowsOf [] = []
    rowsOf ps =
      let (rowPixels, rest) = splitAt w ps
      in renderRow rowPixels : rowsOf rest
    renderRow ps = concatMap pixelToAnsi ps ++ "\ESC[0m"
    pixelToAnsi (Pixel ch r g b) =
      "\ESC[38;2;" ++ show r ++ ";" ++ show g ++ ";" ++ show b ++ "m" ++ [ch]

data RayState = RayState
  { rsTheta       :: !Double
  , rsPhi         :: !Double
  , rsDist        :: !Double
  , rsMorph       :: !Double
  , rsMorphTarget :: !Double
  , rsAutoRotate  :: !Bool
  , rsTick        :: !Int
  }

initialState :: RayState
initialState = RayState
  { rsTheta = 0.0
  , rsPhi = 0.35
  , rsDist = 3.8
  , rsMorph = 0.0
  , rsMorphTarget = 0.0
  , rsAutoRotate = True
  , rsTick = 0
  }

data RayMsg
  = RTick
  | RotL | RotR | RotU | RotD
  | ZoomIn | ZoomOut
  | ToggleAuto
  | NextShape
  deriving (Show, Eq)

-- Render config
fbW, fbH :: Int
fbW = 60
fbH = 28

maxSteps :: Int
maxSteps = 50

maxDist, eps, morphSpeed :: Double
maxDist     = 20.0
eps         = 0.005
morphSpeed  = 0.06

ramp :: String
ramp = " .'`^\",:;Il!i><~+_-?][}{1)(|/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$"

shapeNames :: [String]
shapeNames = ["sphere", "torus", "cube"]

clampD :: Double -> Double -> Double -> Double
clampD lo hi x = max lo (min hi x)

mixD :: Double -> Double -> Double -> Double
mixD a b t = a * (1.0 - t) + b * t

smoothstep :: Double -> Double -> Double -> Double
smoothstep lo hi x =
  let t = clampD 0.0 1.0 ((x - lo) / (hi - lo))
  in t * t * (3.0 - 2.0 * t)

-- Signed-distance functions
sdSphere :: V3 -> Double -> Double
sdSphere p r = vlen p - r

sdTorus :: V3 -> Double -> Double -> Double
sdTorus (V3 x y z) bigR r =
  let qx = sqrt (x * x + z * z) - bigR
  in sqrt (qx * qx + y * y) - r

sdRoundBox :: V3 -> V3 -> Double -> Double
sdRoundBox (V3 px py pz) (V3 bx by bz) r =
  let qx = abs px - bx
      qy = abs py - by
      qz = abs pz - bz
      outer = vlen (V3 (max qx 0) (max qy 0) (max qz 0))
      inner = min (max qx (max qy qz)) 0.0
  in outer + inner - r

scene :: V3 -> Double -> Double
scene p morph =
  let t      = let m = morph - fromIntegral (3 * floor (morph / 3.0) :: Int)
               in if m < 0 then m + 3 else m
      sph    = sdSphere p 1.05
      tor    = sdTorus p 0.9 0.38
      cub    = sdRoundBox p (V3 0.72 0.72 0.72) 0.12
  in if t < 1.0      then mixD sph tor (smoothstep 0 1 t)
     else if t < 2.0 then mixD tor cub (smoothstep 0 1 (t - 1))
     else                 mixD cub sph (smoothstep 0 1 (t - 2))

calcNormal :: V3 -> Double -> V3
calcNormal (V3 x y z) morph =
  let e = 0.001
  in vnorm $ V3
       (scene (V3 (x + e) y z) morph - scene (V3 (x - e) y z) morph)
       (scene (V3 x (y + e) z) morph - scene (V3 x (y - e) z) morph)
       (scene (V3 x y (z + e)) morph - scene (V3 x y (z - e)) morph)

march :: V3 -> V3 -> Double -> Double
march ro rd morph = go 0.0 0
  where
    go !t !i
      | i >= maxSteps || t >= maxDist = -1.0
      | otherwise =
          let d = scene (vadd ro (vscale rd t)) morph
          in if d < eps then t else go (t + d) (i + 1)

lightDir :: V3
lightDir = vnorm (V3 0.8 1.0 (-0.6))

bgPixel :: V3 -> Pixel
bgPixel (V3 _ y _) =
  let vy = y * 0.5 + 0.5
      bg = max 4 (floor (18.0 - (1.0 - vy) * 8.0) :: Int)
  in Pixel ' ' bg bg (bg + 6)

shade :: V3 -> V3 -> Double -> Pixel
shade ro rd morph =
  let t = march ro rd morph
  in if t < 0
       then bgPixel rd
       else
         let hit  = vadd ro (vscale rd t)
             n    = calcNormal hit morph
             diff = max (vdot n lightDir) 0.0
             refl = vsub (vscale n (2.0 * vdot n lightDir)) lightDir
             spec = (max (vdot refl (vneg rd)) 0.0) ** 32.0 * 0.6
             ao   = 1.0 - clampD 0.0 0.4 (scene (vadd hit (vscale n 0.1)) morph * 5.0)
             lum  = clampD 0.0 1.0 ((0.08 + diff * 0.85 + spec) * ao)
             rampLen = length ramp
             idx = floor (clampD 0 (fromIntegral rampLen - 1) (lum * fromIntegral (rampLen - 1))) :: Int
             ch  = ramp !! idx
             V3 nx ny nz = n
             nx' = nx * 0.5 + 0.5
             ny' = ny * 0.5 + 0.5
             nz' = nz * 0.5 + 0.5
             rC  = floor (clampD 0 255 ((nx' * 0.55 + lum * 0.45) * 235 + 20)) :: Int
             gC  = floor (clampD 0 255 ((ny' * 0.45 + lum * 0.55) * 215 + 15)) :: Int
             bC  = floor (clampD 0 255 ((nz' * 0.50 + lum * 0.50 + 0.05) * 200 + 30)) :: Int
         in Pixel ch rC gC bC

renderFrame :: RayState -> L
renderFrame s =
  let theta = rsTheta s
      phi   = rsPhi s
      dist  = rsDist s
      morph = rsMorph s
      ro    = V3 (dist * sin theta * cos phi)
                 (dist * sin phi)
                 (dist * cos theta * cos phi)
      fwd   = vnorm (vneg ro)
      right = vnorm (vcross fwd vup)
      up    = vcross right fwd
      aspect = fromIntegral fbW / fromIntegral fbH * 0.48
      pixels = [ shade ro rd morph
               | py <- [0 .. fbH - 1]
               , let v = 0.5 - fromIntegral py / fromIntegral fbH
               , px <- [0 .. fbW - 1]
               , let u = (fromIntegral px / fromIntegral fbW - 0.5) * aspect
               , let rd = vnorm (vadd (vadd fwd (vscale right u)) (vscale up v))
               ]
  in text (renderFrameBuffer pixels fbW)

-- | Update state in response to a message
updateRay :: RayMsg -> RayState -> RayState
updateRay msg s = case msg of
  RTick ->
    let newTheta = if rsAutoRotate s then rsTheta s + 0.035 else rsTheta s
        diff     = rsMorphTarget s - rsMorph s
        newMorph = if abs diff < 0.01
                     then rsMorphTarget s
                     else rsMorph s + diff * morphSpeed
    in s { rsTheta = newTheta, rsMorph = newMorph, rsTick = rsTick s + 1 }
  RotL  -> s { rsTheta = rsTheta s - 0.15, rsAutoRotate = False }
  RotR  -> s { rsTheta = rsTheta s + 0.15, rsAutoRotate = False }
  RotU  -> s { rsPhi   = min (rsPhi s + 0.1)  1.3 }
  RotD  -> s { rsPhi   = max (rsPhi s - 0.1) (-1.3) }
  ZoomIn  -> s { rsDist = max (rsDist s - 0.25) 2.0 }
  ZoomOut -> s { rsDist = min (rsDist s + 0.25) 8.0 }
  ToggleAuto -> s { rsAutoRotate = not (rsAutoRotate s) }
  NextShape  ->
    let next = (round (rsMorphTarget s) + 1) `mod` 3 :: Int
    in s { rsMorphTarget = fromIntegral next }

-- | Build the visible UI for one frame
viewRay :: RayState -> L
viewRay s =
  let fb        = renderFrame s
      targetIdx = (round (rsMorphTarget s) `mod` 3 + 3) `mod` 3
      settled   = abs (rsMorph s - rsMorphTarget s) < 0.02
      shapeName =
        if settled
          then shapeNames !! targetIdx
          else
            let fromIdx = (round (rsMorph s) `mod` 3 + 3) `mod` 3
            in (shapeNames !! fromIdx) ++ " -> " ++ (shapeNames !! targetIdx)

      sparkData =
        [ 8.0 + 4.0 * sin (fromIntegral (rsTick s + i) * 0.4)
        | i <- [0 .. 29 :: Int]
        ]

      twoPi = 2 * pi :: Double

      cameraStats = withColor ColorBrightBlue $ kv
        [ ("th",   printf "%.2f" (rsTheta s - twoPi * fromIntegral (floor (rsTheta s / twoPi) :: Int)))
        , ("ph",   printf "%.2f" (rsPhi s))
        , ("zoom", printf "%.1f" (rsDist s))
        ]

      rotateLine = tightRow
        [ withColor ColorBrightBlack (text "rotate: ")
        , if rsAutoRotate s
            then withStyle StyleBold (withColor ColorBrightGreen (text "auto"))
            else withColor ColorBrightYellow (text "manual")
        ]

      controls = withStyle StyleDim $ layout
        [ withColor ColorBrightYellow (text "<-^v>  orbit")
        , withColor ColorBrightYellow (text "+/-    zoom")
        , withColor ColorBrightYellow (text "m      shape")
        , withColor ColorBrightYellow (text "a      auto")
        ]

      camera = layout
        [ cameraStats
        , br
        , rotateLine
        , br
        , withColor ColorBrightYellow (text "~12 fps")
        , withColor ColorBrightCyan (plotSparkline sparkData)
        , br
        , withColor ColorBrightMagenta (kv [("shape", shapeName)])
        , br
        , withColor ColorBrightCyan  (spinner "render" (rsTick s `div` 2) SpinnerDots)
        , withColor ColorBrightGreen (spinner "scene"  (rsTick s `div` 3) SpinnerClock)
        , withColor ColorBrightYellow(spinner "light"  (rsTick s `div` 2) SpinnerBounce)
        , br
        , controls
        ]

      cameraBox = withColor ColorBrightMagenta $ setBorder BorderRound $ box "Camera" [camera]

      title = withStyle StyleBold (withColor ColorBrightCyan (text "Ray Marcher"))
  in row [layout [title, fb], cameraBox]

rayMarcher :: LayoutzApp RayState RayMsg
rayMarcher = LayoutzApp
  { appInit = (initialState, CmdNone)
  , appUpdate = \msg state -> (updateRay msg state, CmdNone)
  , appSubscriptions = \_ -> subBatch
      [ subEveryMs 80 RTick
      , subKeyPress $ \case
          KeyLeft      -> Just RotL
          KeyRight     -> Just RotR
          KeyUp        -> Just RotU
          KeyDown      -> Just RotD
          KeyChar '+'  -> Just ZoomIn
          KeyChar '='  -> Just ZoomIn
          KeyChar '-'  -> Just ZoomOut
          KeyChar 'a'  -> Just ToggleAuto
          KeyChar 'A'  -> Just ToggleAuto
          KeyChar 'm'  -> Just NextShape
          KeyChar 'M'  -> Just NextShape
          _            -> Nothing
      ]
  , appView = viewRay
  }

main :: IO ()
main = runApp rayMarcher
